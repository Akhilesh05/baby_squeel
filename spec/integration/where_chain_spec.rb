require 'spec_helper'

describe BabySqueel::ActiveRecord::WhereChain do
  describe '#has' do
    it 'wheres on an attribute' do
      relation = Post.where.has {
        title == 'OJ Simpson'
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        WHERE "posts"."title" = 'OJ Simpson'
      EOSQL
    end

    it 'wheres on associations' do
      relation = Post.joins(:author).where.has {
        author.name == 'Yo Gotti'
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        WHERE "authors"."name" = 'Yo Gotti'
      EOSQL
    end

    it 'wheres using functions' do
      relation = Post.joins(:author).where.has {
        coalesce(title, author.name) == 'meatloaf'
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        WHERE coalesce("posts"."title", "authors"."name") = 'meatloaf'
      EOSQL
    end

    it 'wheres using operations' do
      relation = Post.where.has { (id + 1) == 2 }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        WHERE ("posts"."id" + 1) = 2
      EOSQL
    end

    it 'wheres using complex conditions' do
      relation = Post.joins(:author).where.has {
        (title =~ 'Simp%').or(author.name == 'meatloaf')
      }

      if ActiveRecord::VERSION::STRING < '4.2.0'
        expect(relation).to produce_sql(<<-EOSQL)
          SELECT "posts".* FROM "posts"
          INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
          WHERE (("posts"."title" LIKE 'Simp%' OR "authors"."name" = 'meatloaf'))
        EOSQL
      else
        expect(relation).to produce_sql(<<-EOSQL)
          SELECT "posts".* FROM "posts"
          INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
          WHERE ("posts"."title" LIKE 'Simp%' OR "authors"."name" = 'meatloaf')
        EOSQL
      end
    end

    it 'wheres on associations' do
      relation = Post.joins(author: :comments).where.has {
        author.comments.id > 0
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        INNER JOIN "comments" ON "comments"."author_id" = "authors"."id"
        WHERE ("comments"."id" > 0)
      EOSQL
    end

    it 'wheres on an aliased association' do
      relation = Post.joins(author: :posts).where.has {
        author.posts.id > 0
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        INNER JOIN "posts" "posts_authors" ON "posts_authors"."author_id" = "authors"."id"
        WHERE ("posts_authors"."id" > 0)
      EOSQL
    end

    it 'wheres on an aliased association with through' do
      relation = Post.joins(:comments, :author_comments).where.has {
        author_comments.id > 0
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "comments" ON "comments"."post_id" = "posts"."id"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        INNER JOIN "comments" "author_comments_posts" ON "author_comments_posts"."author_id" = "authors"."id"
        WHERE ("author_comments_posts"."id" > 0)
      EOSQL
    end

    it 'wheres and correctly aliases' do
      relation = Post.joining { author.comments }
                     .where.has { author.comments.id.in [1, 2] }
                     .where.has { author.name == 'Joe' }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        INNER JOIN "comments" ON "comments"."author_id" = "authors"."id"
        WHERE "comments"."id" IN (1, 2) AND "authors"."name" = 'Joe'
      EOSQL
    end

    it 'wheres on an alias with outer join' do
      relation = Post.joining { author.comments.outer }
                     .where.has { author.comments.id.in [1, 2] }
                     .where.has { author.name == 'Joe' }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        LEFT OUTER JOIN "comments" ON "comments"."author_id" = "authors"."id"
        WHERE "comments"."id" IN (1, 2) AND "authors"."name" = 'Joe'
      EOSQL
    end

    it 'wheres on an alias with a function' do
      relation = Post.joins(author: :posts).where.has {
        coalesce(author.posts.id, 1) > 0
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        INNER JOIN "posts" "posts_authors" ON "posts_authors"."author_id" = "authors"."id"
        WHERE (coalesce("posts_authors"."id", 1) > 0)
      EOSQL
    end

    it 'wheres with a subquery' do
      relation = Post.joins(:author).where.has {
        author.id.in Author.selecting { id }.limit(3)
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        WHERE "authors"."id" IN (SELECT "authors"."id" FROM "authors" LIMIT 3)
      EOSQL
    end

    it 'wheres using a simple table' do
      simple = if Arel::VERSION > '7.0.0'
                 BabySqueel[:authors, type_caster: Author.type_caster]
               else
                 BabySqueel[:authors]
               end

      relation = Post.joins(:author).where.has {
        simple.name == 'Yo Gotti'
      }

      expect(relation).to produce_sql(<<-EOSQL)
        SELECT "posts".* FROM "posts"
        INNER JOIN "authors" ON "authors"."id" = "posts"."author_id"
        WHERE "authors"."name" = 'Yo Gotti'
      EOSQL
    end
  end
end
