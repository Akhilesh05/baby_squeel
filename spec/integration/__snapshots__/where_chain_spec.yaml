---
wheres on an attribute 1: SELECT "posts".* FROM "posts" WHERE "posts"."title" = 'OJ
  Simpson'
accepts nil 1: SELECT "posts".* FROM "posts"
wheres on associations 1: SELECT "posts".* FROM "posts" INNER JOIN "authors" ON "authors"."id"
  = "posts"."author_id" WHERE "authors"."name" = 'Yo Gotti'
wheres using functions 1: SELECT "posts".* FROM "posts" INNER JOIN "authors" ON "authors"."id"
  = "posts"."author_id" WHERE coalesce("posts"."title", "authors"."name") = 'meatloaf'
wheres using operations 1: SELECT "posts".* FROM "posts" WHERE ("posts"."id" + 1)
  = 2
wheres using complex conditions 1: SELECT "posts".* FROM "posts" INNER JOIN "authors"
  ON "authors"."id" = "posts"."author_id" WHERE ("posts"."title" LIKE 'Simp%' OR "authors"."name"
  = 'meatloaf')
wheres on an aliased association 1: SELECT "posts".* FROM "posts" INNER JOIN "authors"
  ON "authors"."id" = "posts"."author_id" INNER JOIN "posts" "posts_authors" ON "posts_authors"."author_id"
  = "authors"."id" WHERE ("posts_authors"."id" > 0)
wheres on an aliased association with through 1: SELECT "posts".* FROM "posts" INNER
  JOIN "comments" ON "comments"."post_id" = "posts"."id" INNER JOIN "authors" ON "authors"."id"
  = "posts"."author_id" INNER JOIN "comments" "author_comments_posts" ON "author_comments_posts"."author_id"
  = "authors"."id" WHERE ("author_comments_posts"."id" > 0)
wheres on polymorphic associations 1: SELECT "pictures".* FROM "pictures" INNER JOIN
  "posts" ON "posts"."id" = "pictures"."imageable_id" AND "pictures"."imageable_type"
  = 'Post' WHERE ("posts"."title" LIKE 'meatloaf')
wheres on polymorphic associations outer join 1: SELECT "pictures".* FROM "pictures"
  LEFT OUTER JOIN "posts" ON "posts"."id" = "pictures"."imageable_id" AND "pictures"."imageable_type"
  = 'Post' WHERE ("posts"."title" LIKE 'meatloaf')
wheres and correctly aliases 1: SELECT "posts".* FROM "posts" INNER JOIN "authors"
  ON "authors"."id" = "posts"."author_id" INNER JOIN "comments" ON "comments"."author_id"
  = "authors"."id" WHERE "comments"."id" IN (1, 2) AND "authors"."name" = 'Joe'
wheres on an alias with outer join 1: SELECT "posts".* FROM "posts" INNER JOIN "authors"
  ON "authors"."id" = "posts"."author_id" LEFT OUTER JOIN "comments" ON "comments"."author_id"
  = "authors"."id" WHERE "comments"."id" IN (1, 2) AND "authors"."name" = 'Joe'
wheres on an alias with a function 1: SELECT "posts".* FROM "posts" INNER JOIN "authors"
  ON "authors"."id" = "posts"."author_id" INNER JOIN "posts" "posts_authors" ON "posts_authors"."author_id"
  = "authors"."id" WHERE (coalesce("posts_authors"."id", 1) > 0)
wheres with a subquery 1: SELECT "posts".* FROM "posts" INNER JOIN "authors" ON "authors"."id"
  = "posts"."author_id" WHERE "authors"."id" IN (SELECT  "authors"."id" FROM "authors"
  LIMIT 3)
wheres with an empty subquery 1: SELECT "posts".* FROM "posts" WHERE "posts"."author_id"
  IN (SELECT "authors"."id" FROM "authors" WHERE (1=0))
wheres with an empty subquery and keeps values 1: SELECT "posts".* FROM "posts" WHERE
  "posts"."author_id" IN (SELECT "authors"."id" FROM "authors" INNER JOIN "posts"
  ON "posts"."author_id" = "authors"."id" WHERE (1=0) GROUP BY "authors"."id" ORDER
  BY "authors"."id" ASC)
wheres with a not in subquery 1: SELECT "posts".* FROM "posts" WHERE ("posts"."author_id"
  NOT IN (SELECT "authors"."id" FROM "authors" WHERE (1=0)))
wheres using a simple table 1: SELECT "posts".* FROM "posts" INNER JOIN "authors"
  ON "authors"."id" = "posts"."author_id" WHERE "authors"."name" = 'Yo Gotti'
builds an exists query 1: SELECT "posts".* FROM "posts" WHERE (EXISTS(SELECT "posts".*
  FROM "posts" WHERE "posts"."author_id" = 1))
builds a not exists query 1: SELECT "posts".* FROM "posts" WHERE (NOT EXISTS(SELECT
  "posts".* FROM "posts" WHERE "posts"."author_id" = 1))
'wheres an association using #== 1': SELECT "posts".* FROM "posts" WHERE ("posts"."author_id"
  = 42)
'wheres an association using #!= 1': SELECT "posts".* FROM "posts" WHERE (("posts"."author_id"
  != 42))
